File: src/ObjectManager/Object.Tes/FilePacks/Records/045-TES4.Plugin info.cs
using OA.Core;
using System.Collections.Generic;

namespace OA.Tes.FilePacks.Records
{
    public class TES4Record : Record
    {
        public struct HEDRField
        {
            public float Version;
            public int NumRecords; // Number of records and groups (not including TES4 record itself).
            public uint NextObjectId; // Next available object ID.

            public HEDRField(UnityBinaryReader r, int dataSize)
            {
                Version = r.ReadLESingle();
                NumRecords = r.ReadLEInt32();
                NextObjectId = r.ReadLEUInt32();
            }
        }

        public HEDRField HEDR;
        public STRVField? CNAM; // author (Optional)
        public STRVField? SNAM; // description (Optional)
        public List<STRVField> MASTs; // master
        public List<INTVField> DATAs; // fileSize
        public UNKNField? ONAM; // overrides (Optional)
        public IN32Field INTV; // unknown
        public IN32Field? INCC; // unknown (Optional)

        public override bool CreateField(UnityBinaryReader r, GameFormatId format, string type, int dataSize)
        {
            switch (type)
            {
                case "HEDR": HEDR = new HEDRField(r, dataSize); return true;
                case "OFST": r.SkipBytes(dataSize); return true;
                case "DELE": r.SkipBytes(dataSize); return true;
                case "CNAM": CNAM = r.ReadSTRV(dataSize); return true;
                case "SNAM": SNAM = r.ReadSTRV(dataSize); return true;
                case "MAST": if (MASTs == null) MASTs = new List<STRVField>(); MASTs.Add(r.ReadSTRV(dataSize)); return true;
                case "DATA": if (DATAs == null) DATAs = new List<INTVField>(); DATAs.Add(r.ReadINTV(dataSize)); return true;
                case "ONAM": ONAM = r.ReadUNKN(dataSize); return true;
                case "INTV": INTV = r.ReadT<IN32Field>(dataSize); return true;
                case "INCC": INCC = r.ReadT<IN32Field>(dataSize); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/045-TES4.Plugin info.swift
//
//  TES4Record.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class TES4Record: Record {
    public typealias HEDRField = (
        version: Float,
        numRecords: Int32, // Number of records and groups (not including TES4 record itself).
        nextObjectId: UInt32 // Next available object ID.
    )

    public var HEDR: HEDRField!
    public var CNAM: STRVField? = nil// author (Optional)
    public var SNAM: STRVField? = nil// description (Optional)
    public var MASTs: [STRVField]? = nil// master
    public var DATAs: [INTVField]? = nil// fileSize
    public var ONAM: UNKNField? = nil // overrides (Optional)
    public var INTV: IN32Field! // unknown
    public var INCC: IN32Field? = nil// unknown (Optional)
    // TES5
    public var TNAM: UNKNField? = nil // overrides (Optional)

    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "HEDR": HEDR = r.readT(dataSize)
        case "OFST": r.skipBytes(dataSize)
        case "DELE": r.skipBytes(dataSize)
        case "CNAM": CNAM = r.readSTRV(dataSize)
        case "SNAM": SNAM = r.readSTRV(dataSize)
        case "MAST": if MASTs == nil { MASTs = [STRVField]() }; MASTs!.append(r.readSTRV(dataSize))
        case "DATA": if DATAs == nil { DATAs = [INTVField]() }; DATAs!.append(r.readINTV(dataSize))
        case "ONAM": ONAM = r.readBYTV(dataSize)
        case "INTV": INTV = r.readT(dataSize)
        case "INCC": INCC = r.readT(dataSize)
        // TES5
        case "TNAM": TNAM = r.readBYTV(dataSize)
        default: return false
        }
        return true
    }
}
